Feature/view discover

isaacmartinezimv opened the pull request:
[[FARM-68,](https://amidodevelopment.atlassian.net/secure/RapidBoard.jspa?rapidView=150&projectKey=FARM&view=planning&selectedIssue=FARM-68) [FARM-86](https://amidodevelopment.atlassian.net/secure/RapidBoard.jspa?rapidView=150&projectKey=FARM&view=planning&selectedIssue=FARM-86)]

#### 📲 What

Get list of articles from firestore
Show list of articles
Article list item design
Navigate to article description
Show image and enriched text

#### 🤔 Why

Is important for the users to see a list of articles and to navigate them to read it.
		
#### 🛠 How

Through a list view of cards with rounded borders, with expanded widget for texts.

#### 👀 See

![gif_graphic12](https://user-images.githubusercontent.com/34514354/57357350-045ea500-7173-11e9-811a-0caf0a3e8860.gif)


#### ✅ Acceptance criteria

- [X] Manually tested
- [X] Showcase video / automation test
- [X] Passing all tests
- [ ] Rebased/merged with latest changes from dev/release and re-tested
- [X] Removed TODOs and FIXMEs

#### ⏰ TODO

- [ ] Change the icon to navigate article description

#### 🕵️‍♂️ How to test

Navigate to discover tab, and go into an article description

#### Reviewers

@farmsmart/amido @farmsmart/wamf @emilianowearemobile @maribel4amido

Commit:
[FARM-68] View list of articles load from firebase

Commit:
[FARM-68] Fixed article list UI

Commit:
[FARM-162] Navigate to article detail

Commit:
[FARM-162] Navigation to article detail fixed and working

Commit:
[FARM-159] App bar updated with the same style that myPlot app bar

Commit:
[FARM-163] Article description title style modified

Commit:
Fixed TODO's

emilianowearemobile commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,83 @@
+                        crossAxisAlignment: CrossAxisAlignment.end,
+                        children: <Widget>[
+                          Image.asset(
+                            "assets/icons/profit_loss.png",
```
Check Icons.arrow_forward_ios for the arrow asset :)

isaacmartinezimv replied:
Done! thnks ;)

LeeMatthewHiggins commented:
Don't forget to add reviewers to all PR´s. Min of 2 people should approve before you merge. We can use https://gifox.io. to create nice animated Demos of work.

LeeMatthewHiggins commented:
Also let´s add the Jira links so they are clickable.

emilianowearemobile commented on lib/data/managers/firestore_manager.dart:
```diff
@@ -51,6 +52,45 @@ class FireStoreManager {
+    List<ArticleEntity> articlesEntities;
+
+    // Filters defined by product definition. - ARTICLES
+    var query = Firestore.instance
```
Maybe we can do a method that returns this query, with schema as parameter.
Like prepareFirestoreQuery(schema.ARTICLE)
and prepareFirestoreQuery(schema.CROPS)
so it can be reused :)

maribel4amido replied:
In addition, there are other filters such as the environment and locale that will need to be considered.

isaacmartinezimv replied:
Done! Thnks for the help

emilianowearemobile commented on lib/data/repositories/discover_repository.dart:
```diff
@@ -0,0 +1,24 @@
+import 'package:farmsmart_flutter/data/managers/firestore_manager.dart';
+import 'package:farmsmart_flutter/data/model/article_entity.dart';
+
+class DiscoverRepository {
```
Repository should reference to what data provides...
in this case, it provides Articles... so it should be an ArticlesRepository... 
I suggest that renaming.

isaacmartinezimv replied:
Done!

emilianowearemobile commented on lib/redux/home/home_state.dart:
```diff
@@ -28,7 +28,7 @@ class HomeState {
   HomeState copyWith(
       {LoadingStatus loadingStatus,
-      MyPlotState myPlotState,
+//      MyPlotState myPlotState,
```
This will likely generate merge conflicts as i deleted it on my branch.
Doesnt matters but, we need to keep in mind that this line should disappear :)

emilianowearemobile commented on lib/redux/middleware/discover_middleware.dart:
```diff
@@ -0,0 +1,17 @@
+import 'package:farmsmart_flutter/redux/home/discover/discover_actions.dart';
+import 'package:redux/redux.dart';
+
+class DiscoverMiddleWare extends MiddlewareClass<AppState> {
```
This makes sense to refer it as Discover and not ArticleMiddleware :)

emilianowearemobile replied:
+1

emilianowearemobile commented on lib/redux/middleware/navigation_middleware.dart:
```diff
@@ -13,6 +14,9 @@ class NavigationMiddleware extends MiddlewareClass<AppState>{
       Keys.navKey.currentState.pushNamed(AppRoutes.cropDetail);
     }
+    if(action is GoToArticleDetailAction) {
+      Keys.navKey.currentState.pushNamed(AppRoutes.articleDetail);
```
maybe a switch case is better for this class?

isaacmartinezimv replied:
with switch isn't working for the moment. The middleware seems do not support switch cases yet.

emilianowearemobile commented on lib/ui/app_bar.dart:
```diff
@@ -18,7 +18,8 @@ abstract class CustomAppBar {
       case ARTICLES_TAB:
-        return buildWithTitle(Strings.discoverTab);
+        //return buildWithTitle(Strings.discoverTab);
+        return buildForHome(Strings.discoverTab, profileAction(), homeIcon());
```
Probably, we wont want to have access to the profile from this screen.
But we can ask.

isaacmartinezimv replied:
i put it because is in the UX/UI designs

emilianowearemobile commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,83 @@
+                    Padding(
+                      padding: const EdgeInsets.only(left: 10),
+                    ),
+                    Expanded(
```
I think we should keep, every "child" inside a private method, in the class...
this way you would have like 
Row ( children [
buildTitle,
buildImage,
buildSubtitle
]

all as private members of the class.
I think its a healthy practice, agree?

isaacmartinezimv replied:
Done !

emilianowearemobile reviewed (changes requested):
A bunch of suggestions :)

harrylusk commented:
I've spoken to Emma and it's our preference that the designs be aligned to the UX as per the acceptance criteria in the ticket. Is this possible? Does this require significant rework?

![Screenshot 2019-05-07 at 14 10 10](https://user-images.githubusercontent.com/44875027/57301860-dbd29e80-70d1-11e9-8021-ec9ce9db6c4a.png)

maribel4amido reviewed (commented):
When specifying the JIRA, please ensure that the Sprint JIRA is specified (and not just the subtasks). 

This allows our product owners/ba to easily associate a PR to a user story.

harrylusk reviewed (changes requested):
I've spoken to Emma and it's our preference that the designs be aligned to the UX as per the acceptance criteria in the ticket. Is this possible? Does this require significant rework?

![Screenshot 2019-05-07 at 14 10 10](https://user-images.githubusercontent.com/44875027/57301860-dbd29e80-70d1-11e9-8021-ec9ce9db6c4a.png)

maribel4amido commented on lib/ui/discover/discover_child.dart:
```diff
@@ -1,14 +1,47 @@
+      case LoadingStatus.SUCCESS:
+        return _buildList(context, viewModel.articlesList, viewModel.goToArticleDetail);
+      case LoadingStatus.ERROR:
+        return Text("Error");
```
Add todo as per https://amidodevelopment.atlassian.net/browse/FARM-203

maribel4amido commented on lib/ui/discover/discover_child.dart:
```diff
@@ -1,14 +1,47 @@
 
+Widget _buildList(BuildContext context, List<ArticleEntity> articleList, goToArticleDetail) {
+  return ListView(
+    padding: const EdgeInsets.only(top: 20.0),
```
I suggest placing padding styles in a common file like the TextStyles.

maribel4amido commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,83 @@
+          goToArticleDetail(articleData);
+        },
+        child: Padding(
+            key: ValueKey(articleData.title ?? "No Title"),
```
Place string constants in a common file

maribel4amido commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,83 @@
+            child: Container(
+                height: 90,
+                decoration: BoxDecoration(
+                    color: Colors.white,
```
Please refer to the colors common file or colors class

isaacmartinezimv replied:
that color property has been deleted to approach the view to the deign pattern, but yes it should be refer to our colors common file, Thnks !!

maribel4amido commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,83 @@
+                              maxLines: 4,
+                              overflow: TextOverflow.ellipsis,
+                              style: Styles.footerTextStyle()),
+                          //Margins.generalListSmallMargin(),
```
Please clean unused comments

isaacmartinezimv replied:
Done!

maribel4amido commented on lib/ui/discover/discover_detail_screen.dart:
```diff
@@ -0,0 +1,62 @@
+            automaticallyImplyLeading: true,
+          ),
+          body : Container(
+              decoration: BoxDecoration(
```
Should decorations be defined same as TextStyle?

isaacmartinezimv replied:
that decoration was no necessary, so i just deleted.
We are in process of making every "child" inside a private method, in the class that uses it, so if after that is still necessary to make a common BoxDecorations file, it's no problem at all :)

maribel4amido commented on lib/ui/discover/discover_child.dart:
```diff
@@ -1,14 +1,47 @@
-class HomeDiscoverChild extends StatelessWidget {
-  HomeDiscoverChild();
 
+class HomeDiscoverChild extends StatefulWidget {
```
I don't quite get the name HomeDiscoverChild.  It looks to be the view articles page but I got confused on the use of Home and Child.

maribel4amido replied:
Seems to be the same approach as myplot.

From what I understand: 
child -> same concept as page
detail screen -> the actual detail screen of the child

In my understanding child and detail can be the same thing but it's different here. The child is actually the parent screen.
Is my understanding correct?

albertmiro replied:
I am agree with you @maribel4amido, let's find a better name if this one is making confusion

isaacmartinezimv replied:
Child is referred to the bottomBar, so myPlot, P/L, Discover and Community are every child of the bottom bar.
Item is every element listed.
Detail is the screen inside every item in the list

But you are right with the naming,i think is a little bit confusing so we can rename all the children of bottom bar to Page, like:

 HomeDiscoverChild => HomeDiscoverPage,
 and HomeMyPlotChild => HomeMyPlotPage

Thnks!

maribel4amido replied:
Thanks guys!

albertmiro commented on lib/data/model/article_entity.dart:
```diff
@@ -0,0 +1,35 @@
+    //relatedArticles: articleDocument.data[""], //TODO [FARM-95]
+    status: statusValues.map[articleDocument.data["status"]],
+    summary: articleDocument.data["summary"],
+    title: articleDocument.data["title"]
```
Can we move the string literals to constants?

isaacmartinezimv replied:
for sure, Done!

Commit:
[FARM-158, 159, 160, 162, 163  PR Fixes]   Added the correct Icon for articles List,
File discover_repository renamed to articles_repository,
Class DiscoverRepository renamed to Articles,
MyPlotState line deleted in home_state to prevent merge conflicts with Emi,

Thaiswamf commented:
> I've spoken to Emma and it's our preference that the designs be aligned to the UX as per the acceptance criteria in the ticket. Is this possible? Does this require significant rework?
> 
> ![Screenshot 2019-05-07 at 14 10 10](https://user-images.githubusercontent.com/44875027/57301860-dbd29e80-70d1-11e9-8021-ec9ce9db6c4a.png)



> I've spoken to Emma and it's our preference that the designs be aligned to the UX as per the acceptance criteria in the ticket. Is this possible? Does this require significant rework?
> 
> ![Screenshot 2019-05-07 at 14 10 10](https://user-images.githubusercontent.com/44875027/57301860-dbd29e80-70d1-11e9-8021-ec9ce9db6c4a.png)

Hi @harrylusk, yes, we can change the design removing the rounded corners and the shadow, but the changes to add the line will not be as easy. 
We have two options to do the changes to add the line: Option 1, we can add a line under each article that fill the entire screen (with the same device width) easily on this PR. Option 2, if we want to put the separating line only above the text and we want that it has a width smaller than the size of the screen, it would require some time because it will be needed to change the development of some components on the screen and these tasks would be needed to be defined as a new task for the next sprint.
What option do you prefer to follow for this screen? Thank you

jessicahernandezhv commented:
Hello @harrylusk, yesterday we've been investigating about the Option 2 that @Thaiswamf proposed (make a separator only above the text, as the UI design shows).

After a deep research, we just found an option to make it, changing the development of the minimum number of components, that make possible to include it in this PR. Thank you.

harrylusk commented:
@jessicahernandezhv @Thaiswamf thanks for your time researching this. If option 2 is simple, let's go with that. Otherwise, I'm happy if the line displays across the full width of the screen.

harrylusk commented:
@jessicahernandezhv also, can we change the icon from the coins to the arrow?

albertmiro commented on lib/ui/app_bar.dart:
```diff
@@ -18,7 +18,8 @@ abstract class CustomAppBar {
         break;
       case ARTICLES_TAB:
-        return buildWithTitle(Strings.discoverTab);
+        //return buildWithTitle(Strings.discoverTab);
```
If we are not going to use it, let's remove the comment

isaacmartinezimv replied:
Done!

albertmiro commented on lib/ui/discover/discover_child.dart:
```diff
@@ -1,14 +1,47 @@
+      case LoadingStatus.SUCCESS:
+        return _buildList(context, viewModel.articlesList, viewModel.goToArticleDetail);
+      case LoadingStatus.ERROR:
+        return Text("Error");
```
` return Text("Error");`
Let's move this string to a const value, or some place where we can localize later on.

isaacmartinezimv replied:
Done!

albertmiro commented on lib/ui/discover/discover_child_item.dart:
```diff
@@ -0,0 +1,72 @@
+                      child: Column(
+                        crossAxisAlignment: CrossAxisAlignment.start,
+                        children: <Widget>[
+                          Text(articleData.title ?? "No Title",
```
Move the string to a common strings file, we shouldn't have hardcoded values in the app

isaacmartinezimv replied:
Done!

albertmiro commented on lib/ui/discover/discover_viewmodel.dart:
```diff
@@ -0,0 +1,32 @@
+      loadingStatus:  store.state.discoverState.loadingStatus,
+      articlesList: store.state.discoverState.articleList,
+      selectedArticle: store.state.discoverState.selectedArticle,
+      goToArticleDetail: (ArticleEntity article) => store.dispatch(GoToArticleDetailAction(article)),
```
Maybe we can add the `store.dispatch(GoToArticleDetailAction(article))` to a new line to follow the same as the next line with the fetchArticles

isaacmartinezimv replied:
totally agree ! it keeps the code more easy to read.

albertmiro commented on lib/ui/discover/discover_viewmodel.dart:
```diff
@@ -0,0 +1,32 @@
+      selectedArticle: store.state.discoverState.selectedArticle,
+      goToArticleDetail: (ArticleEntity article) => store.dispatch(GoToArticleDetailAction(article)),
+      fetchArticles: () =>
+          store.dispatch(new FetchArticleListAction())
```
`new` is no longer needed in Dart 2

isaacmartinezimv replied:
yes, sometimes i forgot that, it's removed now, thnks !

albertmiro commented on lib/ui/discover/discover_detail_screen.dart:
```diff
@@ -0,0 +1,62 @@
+                        Margins.generalListMargin(),
+                        Html(data: selectedArticleData.content),
+                      ],
+                    )),
```
Can we move the Widget to a method that can return it, and then another one to return the Padding?

We would have something like, I think is more readable:
```
  child: ListView(children: <Widget>[
                getFadeImage(selectedArticleData),
                getPadding(selectedArticleData),
              ]))
```

isaacmartinezimv replied:
Done !

emilianowearemobile replied:
Albert, don't copy my comments! xD

albertmiro replied:
@emilianowearemobile I haven't seen yours! <image src="https://user-images.githubusercontent.com/15966505/57441666-b0bf8a80-724a-11e9-99f0-358d240eb456.png" width=8%/>

isaacmartinezimv commented:
> @jessicahernandezhv also, can we change the icon from the coins to the arrow?

yeah, its changed, thnks Harry !

Commit:
[Peer comments] article_entity strings moved into literals file Strings.dart, removed the unused line //return buildWithTitle(Strings.discoverTab); in app_bar,
String "Error" in discover_child moved into Strings file,
store.dispatch(GoToArticleDetailAction(article)) from discover_viewmodel moved to a nel line,
"new" statement deleted from discover_detail_screen.

Commit:
[PR Comments] my_plot_child renamed into my_plot_page, discover_child renamed into discover_page, HomeMyPlotChild renamed into HomeMyPlotPage,
HomeDiscoverChild renamed into HomeDiscoverPage, all "child" components in discover_page moved into private methods to clarify the code.

emilianowearemobile commented on lib/utils/strings.dart:
```diff
@@ -27,5 +30,11 @@ class Strings {
   static String myPlotDetailCompanionPlantsTitle = "Companion Plants";
   static String myPlotDetailNonCompanionPlantsTitle = "Non-Companion Plants";
+    
+  static String documentFieldContent = "content";
```
These are not "UI Strings" that are going to be localized neither shown to the user.
They are constants for firebase, check firebase_const file :)

tenmaster committed:
[MERGE] Update merge conflicts

tenmaster committed:
[PR COMMENTS] Move firebase strings to const file.

Commit:
[PR Comments] Image size put to .cover in the article_list, app_bar on article_detail_screen now it's like the app_bar on crop_detail page, padding between elements removed on articles list.
